contributors/design-proposals/auth/node-tls-bootstrap-via-tpm.md

```diff
@@ -0,0 +1,280 @@
+design.
+
+#### Key pair stored in TPM
+
+TPMs can generate and store key pairs and make them non-exportable: the private
+key never leaves TPM API boundary.
+[tls.Certificate.PrivateKey](https://golang.org/pkg/crypto/tls/#Certificate)
+allows replacing in-memory key with a custom crypto.Signer implementation.
+
+#### Certificate rotation
+
+Certificate rotation currently uses existing certificate to request a new one.
+If existing certificate expires before rotation succeeds (e.g. when node is
+partitioned from cluster or suspended), the node requires operator
+intervention.  Rotation can be changed to use the TPM-based flow, eliminating
+the need for manual intervention.
+
+#### Incorporate PCR quoting
+
+PCR quotes can be added to the CSR. These quotes record the state of the
+machine (firmware, kernel parameters, etc). Master may validate these values
+during CSR approval.
+
+## Implementation options
+
+### Option 1: directly in Kubelet
+
+#### New package `tpm`
+
+Pckage `tpm` will have helper funcs for TPM-related functionality. Uses
+[github.com/google/go-tpm/tpm2](https://github.com/google/go-tpm/tree/master/tpm2)
+under the hood.
+
+```go
+package tpm
+
+type TPM struct {...}
+func Open(path) (*TPM, error)
+func (t *TPM) Close() error
+
+// AttestKey returns series of PEM blocks with attestation data.
+func (t *TPM) AttestKey(privateKey) (attestData, error)
+// VerifyCSR parses CSR PEM blocks with attestation data and verifies the
+// signature (used by Approver).
+func (t *TPM) VerifyCSR(csrData) error
+```
+
+#### New kubeconfig fields
+
+New fields for TPM interaction are added under `users` in `kubeconfig`
+([AuthInfo
+struct](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/api/types.go)):
+
+1. BootstrapCSRAttestation={none|tpm} - extra attestation mode for CSRs. none
+   is default. Field is behind a [feature
+   gate](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-field-in-existing-api-version).
+   Note: other values may be added in the future for alternative attestation
+   schemes.
+1. TPM (new type TPMConfig) - TPM-specific config fields
+1. TPM.Path string - path to TPM device (character device or unix socket)
+1. TPM.AIKCertIndex int - NVRAM index of AIK certificate
+1. TPM.AIKTemplateIndex int - NVRAM index of AIK template
+
+#### Kubelet code
+
+[bootstrap.LoadClientCert](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/certificate/bootstrap/bootstrap.go#L47)
+is the entry point.
+
+Split [csr.RequestNodeCertificate](https://github.com/kubernetes/client-go/blob/master/util/certificate/csr/csr.go#L51) into
+
+```go
+func MakeNodeCSR(privateKey, nodeName) (csrData, error)
+func RequestNodeCertificate(client, csrData, privateKey) (certData, error)
+```
+
+In `bootstap.LoadClientCert`:
+
+```
+csrData, err := csr.MakeNodeCSR(privateKey, nodeName)
+// handle err
+switch config.BootstrapCSRAttestation {
```

Comment:
Lots of cloud based instance identity strategies will want to inject themselves here too. I'm skeptical of adding more hard coded hooks like this into the codebase. Maybe others can weigh in here?